dev/src/bundle.ts

```diff
@@ -0,0 +1,225 @@
+      this.querySnaps.set(docOrName, query);
+    }
+
+    return this;
+  }
+
+  private addBundledDocument(snap: DocumentSnapshot) {
+    if (snap.exists) {
+      const docProto = snap.toDocumentProto();
+      if (
+        !this.documents.has(snap.id) ||
+        Timestamp.fromProto(this.documents.get(snap.id)!.metadata.readTime!) <
+          snap.readTime
+      ) {
+        this.documents.set(snap.id, {
+          document: docProto,
+          metadata: {
+            documentKey: docProto.name,
+            readTime: snap.readTime.toProto().timestampValue,
+          },
+        });
+      }
+      if (snap.readTime > this.latestReadTime) {
+        this.latestReadTime = snap.readTime;
+      }
+    }
+  }
+
+  private addNamedQuery(name: string, querySnap: QuerySnapshot) {
+    this.namedQueries.set(name, {
+      name,
+      bundledQuery: querySnap.query.toBundledQuery(),
+      readTime: querySnap.readTime.toProto().timestampValue,
+    });
+
+    for (const snap of querySnap.docs) {
+      this.addBundledDocument(snap);
+    }
+
+    if (querySnap.readTime > this.latestReadTime) {
+      this.latestReadTime = querySnap.readTime;
+    }
+  }
+
+  private pushLengthPrefixedString(readable: Readable, payload: string) {
+    const buffer = Buffer.from(payload, 'utf-8');
+    const lengthBuffer = Buffer.from(buffer.length.toString());
+
+    readable.push(Buffer.concat([lengthBuffer, buffer]));
+  }
+
+  stream(): NodeJS.ReadableStream {
+    assert(
+      this.docRefs.length +
+        this.docSnaps.length +
+        this.queries.size +
+        this.querySnaps.size >
+        0,
+      'Nothing is added to the bundle.'
+    );
+    for (const snap of this.docSnaps) {
+      this.addBundledDocument(snap);
+    }
+
+    for (const [name, snap] of Array.from(this.querySnaps)) {
+      if (this.namedQueries.has(name)) {
+        throw new Error(`Query name conflict: ${name}`);
+      }
+
+      this.addNamedQuery(name, snap);
+    }
+
+    const promises: Array<Promise<void>> = [];
+    for (const doc of this.docRefs) {
+      promises.push(
+        doc.get().then(snap => {
+          this.addBundledDocument(snap);
+        })
+      );
+    }
+
+    for (const [name, query] of Array.from(this.queries)) {
+      if (this.namedQueries.has(name)) {
+        throw new Error(`Query name conflict: ${name}`);
+      }
+
+      promises.push(
+        query.get().then(snap => {
+          const querySnap = snap as QuerySnapshot;
+          this.addNamedQuery(name, querySnap);
+        })
+      );
+    }
+
+    const readable = new Readable({
+      objectMode: false,
+      read: async size => {
+        await Promise.all(promises);
+        const metadata: firestore.IBundleMetadata = {
+          id: this.id,
+          createTime: this.latestReadTime.toProto().timestampValue,
+        };
+        this.pushLengthPrefixedString(
+          readable,
+          JSON.stringify({metadata} as firestore.IBundleElement)
```

Comment:
The cast is not valid (since `metadata` is `IBundleMetadata`, but not `{metadata}`). You can just drop the cast.

Reply:
I think it is valid but unnecessary, because it is casting to `IBundleElement`.

I slightly prefer to keep these though, because it can serve as a reminder that it is a `IBundleElement` that gets saved in the bundle.